### PR TITLE
Show correct errors when plugins yaml forgot the 'flutter.plugins.pla…

### LINKED ISSUE #61337: [flutter_tool] flutter plugin yaml format not correct but validation let it pass
If a plugin is migrating from old plugin format to new plugin format (The multi-platform format) and forgot to add the key `flutter.plugin.platforms` at `pubspec.yaml` like below

```yaml
flutter:
  plugin:
    android:
      package: packageName
      pluginClass: className
```

and users upgrade the plugin, they will not receive any error but the build will pass as before.

BUT the plugin's platform side registrar already missing, because the `flutter_tool` can't found the correct platform type for it to generate it.

I was meet this problem recently, I will create a PR associate to this issue
  
> https://pub.dev/packages/flutter_growingio_track  version: 2.6.5  is a real case

### PR DESCRIPTION
**Pull request by Chaoyang0201**
## Description

If a plugin is migrating from old plugin format to new plugin format (The multi-platform format) and forgot to add the key `flutter.plugin.platforms` at `pubspec.yaml` like below

```yaml
flutter:
  plugin:
    android:
      package: packageName
      pluginClass: className
```

and users upgrade the plugin, they will not receive any error but the build will pass as before.

BUT the plugin's platform side registrar already missing, because the `flutter_tool` can't found the correct platform type for it to generate it.

I was meet this problem recently, so create this MR to avoid this problem again.

> https://pub.dev/packages/flutter_growingio_track  version: 2.6.5  is a real case 

This is my first time to contribute to flutter,  if has any issues of this MR, please let me know.

## Related Issues

Fixes #61337

## Tests

I added the following tests:
```dart
    testWithoutContext('FlutterManifest validates missing flutter.plugin.platforms key', () {
    const String manifest = '''
name: test
flutter:
  plugin:
    android:
      package: com.example
      pluginClass: SomeClass
    ios:
      pluginClass: SomeClass
''';
    final BufferLogger logger = BufferLogger.test();
    final FlutterManifest flutterManifest = FlutterManifest.createFromString(
      manifest,
      logger: logger,
    );

    expect(flutterManifest, null);
    expect(logger.errorText,
      contains('The flutter.plugin.platforms key should be added,'
      'when specifying a plugin supported platforms'));
  });
}
```

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/plugins.dart [hunk 1]
```diff
@@ -264,6 +264,19 @@ class Plugin {
 
   static List<String> _validateLegacyYaml(YamlMap yaml) {
     final List<String> errors = <String>[];
+
```
Comment (cyanglaz) [inline packages/flutter_tools/lib/src/plugins.dart:267]: I feel like it doesn't make sense to put these checks here as the method name says "_validateLegacyYaml", but the code is actually validating the new format.

What do you think about fixing it in the `validatePluginYaml`? We have something like:

```dart
if (usesNewPluginFormat) {
  if (yaml['platforms'] != null && yaml['platforms'] is! YamlMap) {
    const String errorMessage = 'flutter.plugin.platforms should be a map with the platform name as the key';
    return <String>[errorMessage];
  }
  return _validateMultiPlatformYaml(yaml['platforms'] as YamlMap);
} else {
  return _validateLegacyYaml(yaml);
}
```

We can make it as:

```dart
if (usesNewPluginFormat) {
  if (yaml['platforms'] != null && yaml['platforms'] is! YamlMap) {
    const String errorMessage = 'flutter.plugin.platforms should be a map with the platform name as the key';
    return <String>[errorMessage];
  }
  return _validateMultiPlatformYaml(yaml['platforms'] as YamlMap);
} else if (usesOldPluginFormat) {
  return _validateLegacyYaml(yaml);
} else {
  return <String>['The flutter.plugin.platforms key in the `pubspec.yaml` is not found. '
                              'An instruction to format the `pubspec.yaml` can be found here: ' 
                              'https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin'];
}
```

File: packages/flutter_tools/test/general.shard/flutter_manifest_test.dart [hunk 2]
```diff
@@ -963,6 +963,29 @@ flutter:
     expect(logger.errorText,
       contains('flutter.plugin.platforms should be a map with the platform name as the key'));
   });
+
+    testWithoutContext('FlutterManifest validates missing flutter.plugin.platforms key', () {
```
Comment (cyanglaz) [inline packages/flutter_tools/test/general.shard/flutter_manifest_test.dart:967]: This is great! Thanks :)

File: packages/flutter_tools/lib/src/plugins.dart (hunk 1 above)
Comment (Chaoyang0201) [reply packages/flutter_tools/lib/src/plugins.dart:267]: Yeah, put the check inside the method `_validateLegacyYaml` is weird.
I will modify it later, thanks for the suggestion : ) !

Comment (cyanglaz) [reply packages/flutter_tools/lib/src/plugins.dart:275]: Thanks! Feel free to @ me when it's ready to be reviewed again :)

**Comment by Chaoyang0201** [comment]
@cyanglaz     it's ready for review  : )

File: packages/flutter_tools/lib/src/plugins.dart [hunk 3]
```diff
@@ -213,6 +213,14 @@ class Plugin {
       return <String>[errorMessage];
     }
 
+    if (!usesOldPluginFormat && !usesNewPluginFormat) {
+      const String errorMessage =
+          'The flutter.plugin section format at the `pubspec.yaml` is not correct. '
```
Comment (cyanglaz) [inline packages/flutter_tools/lib/src/plugins.dart:218]: What about: 
```
Cannot find the `flutter.plugin.platforms` key in the `pubspec.yaml` file.
```

We want to push users to use the new format.

File: packages/flutter_tools/lib/src/plugins.dart [hunk 4]
```diff
@@ -213,6 +213,14 @@ class Plugin {
...
+          'An instruction to format the `pubspec.yaml` can be found here: '
+          'See: https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms';
```
Comment (cyanglaz) [inline packages/flutter_tools/lib/src/plugins.dart:220]: We don't need the "See:" here.

File: packages/flutter_tools/lib/src/plugins.dart (hunk 3 above)
Comment (Chaoyang0201) [reply packages/flutter_tools/lib/src/plugins.dart:218]: OK.

File: packages/flutter_tools/lib/src/plugins.dart (hunk 4 above)
Comment (Chaoyang0201) [reply packages/flutter_tools/lib/src/plugins.dart:220]: oops

**Comment by Chaoyang0201** [comment]
@cyanglaz   All comments are resolved.

**Comment by Chaoyang0201** [comment]
@cyanglaz   the check `
deploy_gallery-linux`  seems stuck?
it already running 2d 8h 😂

**Review by cyanglaz — APPROVED** [review]
LGTM! I can't seem to re-run the task. @Chaoyang0201 Maybe push an empty commit to re-trigger the ci? 

cc @godofredoc

**Comment by Chaoyang0201** [comment]
> LGTM! I can't seem to re-run the task. @Chaoyang0201 Maybe push an empty commit to re-trigger the ci?
> 
> cc @godofredoc

I pushed to re-trigger it 🙏

**Pull request merged**